File: backend/internal/handler/settings_handler.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/kenyamamoto/pokelingual/backend/internal/domain"
	"github.com/kenyamamoto/pokelingual/backend/internal/service"
)

// SettingsHandler handles HTTP requests for user settings.
type SettingsHandler struct {
	settingsRepo domain.UserSettingsRepository
}

// NewSettingsHandler creates a new SettingsHandler.
func NewSettingsHandler(settingsRepo domain.UserSettingsRepository) *SettingsHandler {
	return &SettingsHandler{
		settingsRepo: settingsRepo,
	}
}

// GetSettings handles GET /settings to retrieve user settings.
func (h *SettingsHandler) GetSettings(c *gin.Context) {
	uid := c.GetString("uid")

	settings, err := h.settingsRepo.GetSettings(c.Request.Context(), uid)
	if err != nil {
		handleError(c, err)
		return
	}

	excluded := settings.ExcludedPokemonIDs
	if excluded == nil {
		excluded = service.DefaultExcludedPokemonIDs
	}

	c.JSON(http.StatusOK, gin.H{
		"excluded_pokemon_ids": excluded,
		"max_pokemon_id":      service.MaxPokemonID,
	})
}

type updateExcludedPokemonRequest struct {
	PokemonIDs []int `json:"pokemon_ids" binding:"required"`
}

// UpdateExcludedPokemon handles PUT /settings/excluded-pokemon.
func (h *SettingsHandler) UpdateExcludedPokemon(c *gin.Context) {
	uid := c.GetString("uid")

	var req updateExcludedPokemonRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
		return
	}

	if err := h.settingsRepo.UpdateExcludedPokemon(c.Request.Context(), uid, req.PokemonIDs); err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{"status": "ok"})
}

File: backend/src/handler/settings-handler.ts
import type { Request, Response } from "express";
import type { SettingsService } from "../service/settings-service.js";
import type {
  UpdateExcludedPokemonRequest,
  UpdateEnabledGenerationsRequest,
} from "../../../shared/api-types/settings.js";
import type { ErrorResponse } from "../../../shared/api-types/error.js";
import { handleError } from "./error.js";

/** ユーザ設定 (除外ポケモン等) のエンドポイントを束ねるハンドラ。 */
export class SettingsHandler {
  /**
   * @param settingsService ユーザ設定のドメインサービス。
   */
  constructor(private settingsService: SettingsService) {}

  /**
   * GET /settings — ユーザ自身の除外ポケモンIDを返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  getSettings = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    try {
      const body = await this.settingsService.getSettings(userId);
      res.json(body);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * PUT /settings/excluded-pokemon — 除外ポケモンIDリストをバリデーションして更新する。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  updateExcludedPokemon = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    const body = req.body as Partial<UpdateExcludedPokemonRequest>;
    try {
      const result = await this.settingsService.updateExcludedPokemon(userId, body.pokemon_ids);
      if (!result.ok) {
        res.status(400).json({ error: result.message } satisfies ErrorResponse);
        return;
      }
      res.json({ status: "ok" });
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * PUT /settings/generations — 出題対象の世代リストをバリデーションして更新する。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  updateEnabledGenerations = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    const body = req.body as Partial<UpdateEnabledGenerationsRequest>;
    try {
      const result = await this.settingsService.updateEnabledGenerations(userId, body.generations);
      if (!result.ok) {
        res.status(400).json({ error: result.message } satisfies ErrorResponse);
        return;
      }
      res.json({ status: "ok" });
    } catch (err) {
      handleError(res, err, req.path);
    }
  };
}
